examples/fwi_operators.py

```diff
@@ -1,9 +1,135 @@
 from devito.operators import *
 from sympy import Eq
-from devito.interfaces import TimeData, DenseData
+from devito.interfaces import TimeData, DenseData, PointData
 from sympy import Function, symbols, as_finite_diff, Wild, IndexedBase
 from sympy.abc import x, y, t, z
-from sympy import solve
+from sympy import solve, Matrix
+
+
+class SourceLike(PointData):
+    """Defines the behaviour of sources and receivers.
+    """
+    def __init__(self, name, npoint, nt, dt, h, data, ndim, dtype, nbpml):
+        self.orig_data = data
+        self.dt = dt
+        self.h = h
+        self.ndim = ndim
+        self.nbpml = nbpml
+        super(SourceLike, self).__init__(name, npoint, nt, dtype)
+        x1, y1, z1, x2, y2, z2 = symbols('x1, y1, z1, x2, y2, z2')
+        if ndim == 2:
+            A = Matrix([[1, x1, z1, x1*z1],
+                        [1, x1, z2, x1*z2],
+                        [1, x2, z1, x2*z1],
+                        [1, x2, z2, x2*z2]])
+            self.increments = (0, 0), (0, 1), (1, 0), (1, 1)
+            self.rs = symbols('rx, rz')
+            rx, rz = self.rs
+            p = Matrix([[1],
+                        [rx],
+                        [rz],
+                        [rx*rz]])
+        else:
+            A = Matrix([[1, x1, y1, z1, x1*y1, x1*z1, y1*z1, x1*y1*z1],
+                        [1, x1, y2, z1, x1*y2, x1*z1, y2*z1, x1*y2*z1],
+                        [1, x2, y1, z1, x2*y1, x2*z1, y2*z1, x2*y1*z1],
+                        [1, x1, y1, z2, x1*y1, x1*z2, y1*z2, x1*y1*z2],
+                        [1, x2, y2, z1, x2*y2, x2*z1, y2*z1, x2*y2*z1],
+                        [1, x1, y2, z2, x1*y2, x1*z2, y2*z2, x1*y2*z2],
+                        [1, x2, y1, z2, x2*y1, x2*z2, y1*z2, x2*y1*z2],
+                        [1, x2, y2, z2, x2*y2, x2*z2, y2*z2, x2*y2*z2]])
+            self.increments = (0, 0, 0), (0, 1, 0), (1, 0, 0), (0, 0, 1), (1, 1, 0), (0, 1, 1), (1, 0, 1), (1, 1, 1)
+            self.rs = symbols('rx, ry, rz')
+            rx, ry, rz = self.rs
+            p = Matrix([[1],
+                        [rx],
+                        [ry],
+                        [rz],
+                        [rx*ry],
+                        [rx*rz],
+                        [ry*rz],
+                        [rx*ry*rz]])
+
+        # Map to reference cell
+        reference_cell = [(x1, 0),
+                          (y1, 0),
+                          (z1, 0),
+                          (x2, self.h),
+                          (y2, self.h),
+                          (z2, self.h)]
+        A = A.subs(reference_cell)
+        self.bs = A.inv().T.dot(p)
+
+    def point2grid(self, pt_coords):
```

Comment:
Yes, just a brief statement of "what" alongside the existing comments on the expected input and output arguments should be fine.